In [1]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',
optimizer=OPTIMIZER,
metrics=['accuracy'])
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])




Using TensorFlow backend.


60000 train samples
10000 test samples
Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_1 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_2 (Activation)    (None, 128)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 10)                1290      
_________________________________________________________________
activation_3 (Activation)    (None, 10)                0         
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________

In [2]:
# from Deep learing with Keras Pages 25-26
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
DROPOUT = 0.3
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers 10 outputs
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',
optimizer=OPTIMIZER,
metrics=['accuracy'])
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])

Model: "sequential_2"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_4 (Dense)              (None, 128)               100480    
_________________________________________________________________
activation_4 (Activation)    (None, 128)               0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_5 (Dense)              (None, 128)               16512     
_________________________________________________________________
activation_5 (Activation)    (None, 128)               0         
_________________________________________________________________
dropout_2 (Dropout)          (None, 128)               0         
_________________________________________________________________
dense_6 (Dense)              (None, 10)               

In [1]:
# Aaron Cimielli - March 11 2023 
# Hidden Layers Experiment
# For this experiment I modified the Number of hidden layers
# N_HIDDEN [1, 5, 20] 
# After each model we printed I printed the number of hidden layers and the test accuracy 
# 

# Examples were found in chapter 1 and 2 from Deep learing with Keras
# and https://keras.io

from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 20
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = [1, 5, 20]
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers
# 10 outputs
# final stage is softmax
for i in range(len(N_HIDDEN)):
    model = Sequential()
    model.add(Dense(N_HIDDEN[i], input_shape=(RESHAPED,)))
    model.add(Activation('relu'))
    for j in range(N_HIDDEN[i]-1):
        model.add(Dense(N_HIDDEN[i]))
        model.add(Activation('relu'))
    model.add(Dense(NB_CLASSES))
    model.add(Activation('softmax'))
    model.summary()
    model.compile(loss='categorical_crossentropy',
    optimizer=OPTIMIZER,
    metrics=['accuracy'])
    history = model.fit(X_train, Y_train,
    batch_size=BATCH_SIZE, epochs=NB_EPOCH,
    verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
    score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
    print(f"Number of Hidden Layers: {N_HIDDEN[i]}")
    print("Test score:", score[0])
    print('Test accuracy:', score[1])


Using TensorFlow backend.


11493376/11490434 [==============================] - 0s 0us/step
60000 train samples
10000 test samples
Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 1)                 785       
_________________________________________________________________
activation_1 (Activation)    (None, 1)                 0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                20        
_________________________________________________________________
activation_2 (Activation)    (None, 10)                0         
Total params: 805
Trainable params: 805
Non-trainable params: 0
_________________________________________________________________
Train on 48000 samples, validate on 12000 samples
Epoch 1/20
48000/48000 [==============================] - 2s 42us/step - loss: 2.2631 - accuracy: 0.

48000/48000 [==============================] - 2s 35us/step - loss: 0.8271 - accuracy: 0.7603 - val_loss: 0.7872 - val_accuracy: 0.7736
Epoch 15/20
48000/48000 [==============================] - 2s 34us/step - loss: 0.7769 - accuracy: 0.7819 - val_loss: 0.7422 - val_accuracy: 0.7920
Epoch 16/20
48000/48000 [==============================] - 2s 33us/step - loss: 0.7371 - accuracy: 0.7946 - val_loss: 0.7127 - val_accuracy: 0.8010
Epoch 17/20
48000/48000 [==============================] - 2s 33us/step - loss: 0.7039 - accuracy: 0.8051 - val_loss: 0.6781 - val_accuracy: 0.8140
Epoch 18/20
48000/48000 [==============================] - 2s 35us/step - loss: 0.6748 - accuracy: 0.8152 - val_loss: 0.6503 - val_accuracy: 0.8173
Epoch 19/20
48000/48000 [==============================] - 2s 34us/step - loss: 0.6488 - accuracy: 0.8224 - val_loss: 0.6378 - val_accuracy: 0.8256
Epoch 20/20
10000/10000 [==============================] - 0s 18us/step
Number of Hidden Layers: 5
Test score: 0.60479497609

48000/48000 [==============================] - 3s 61us/step - loss: 2.0148 - accuracy: 0.2145 - val_loss: 1.9291 - val_accuracy: 0.2123
Epoch 9/20
48000/48000 [==============================] - 3s 60us/step - loss: 1.8884 - accuracy: 0.2188 - val_loss: 1.8396 - val_accuracy: 0.2176
Epoch 10/20
48000/48000 [==============================] - 3s 61us/step - loss: 1.8153 - accuracy: 0.2255 - val_loss: 1.8286 - val_accuracy: 0.2321
Epoch 11/20
48000/48000 [==============================] - 3s 64us/step - loss: 1.7641 - accuracy: 0.2403 - val_loss: 1.7313 - val_accuracy: 0.2589
Epoch 12/20
48000/48000 [==============================] - 3s 63us/step - loss: 1.7210 - accuracy: 0.2742 - val_loss: 1.8199 - val_accuracy: 0.2227
Epoch 13/20
48000/48000 [==============================] - 3s 61us/step - loss: 1.6315 - accuracy: 0.3192 - val_loss: 1.4896 - val_accuracy: 0.3833
Epoch 14/20
48000/48000 [==============================] - 3s 65us/step - loss: 1.4820 - accuracy: 0.3809 - val_loss: 1.3530 

In [5]:
# Aaron Cimielli - March 11 2023 
# Batch Size 
# for this experiment I modified the batch size.
# Specifically, I decided to use 3 different values, "32, 64, and 256". 
# Here I am using two hidden layers. 

# Examples were found in chapter 1 and 2 from Deep learing with Keras
# and from https://keras.io/api/models/model_training_apis ( site found in chapter 1 )

from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils

np.random.seed(1671) # for reproducibility

# network and training
NB_EPOCH = 20
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION

# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784

#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

# normalize
X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)

# 2 hidden layers
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()

model.compile(loss='categorical_crossentropy', optimizer=OPTIMIZER, metrics=['accuracy'])

# Experiment 2: Change batch size
batch_sizes = [32, 64, 256]

for batch_size in batch_sizes:
    print(f"\nExperiment with batch size of {batch_size}")
    history = model.fit(X_train, Y_train,
                        batch_size=batch_size, epochs=NB_EPOCH,
                        verbose=1, validation_split=VALIDATION_SPLIT)

    score = model.evaluate(X_test, Y_test, verbose=0)
    print("Test score:", score[0])
    print('Test accuracy:', score[1])


60000 train samples
10000 test samples
Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_11 (Dense)             (None, 128)               100480    
_________________________________________________________________
activation_11 (Activation)   (None, 128)               0         
_________________________________________________________________
dense_12 (Dense)             (None, 128)               16512     
_________________________________________________________________
activation_12 (Activation)   (None, 128)               0         
_________________________________________________________________
dense_13 (Dense)             (None, 10)                1290      
_________________________________________________________________
activation_13 (Activation)   (None, 10)                0         
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________

In [6]:
# Aaron Cimielli - March 11 2023 
# Number of epochs
# for this experiment I modified the number of epochs.
# Specifically, I decided to use 3 different values, "10, 30, 50". 
# Here I am using two hidden layers. 

# Examples were found in chapter 1 and 2 from Deep learing with Keras
# and from https://keras.io/guides/training_with_built_in_methods ( site found in chapter 1 )

from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Activation
from keras.optimizers import SGD
from keras.utils import np_utils

np.random.seed(1671) # for reproducibility

# network and training
BATCH_SIZE = 128
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION

# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784

#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')

# normalize
X_train /= 255
X_test /= 255

print(X_train.shape[0], 'train samples')
print(X_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)

# 2 hidden layers
# 10 outputs
# final stage is softmax
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()

model.compile(loss='categorical_crossentropy',
              optimizer=OPTIMIZER,
              metrics=['accuracy'])

# Experiment 3: Change number of epochs
nb_epochs = [10, 30, 50]

for nb_epoch in nb_epochs:
    print(f"\nExperiment with {nb_epoch} epochs")
    history = model.fit(X_train, Y_train,
                        batch_size=BATCH_SIZE, epochs=nb_epoch,
                        verbose=1, validation_split=VALIDATION_SPLIT)

    score = model.evaluate(X_test, Y_test, verbose=0)
    print("Test score:", score[0])
    print('Test accuracy:', score[1])



60000 train samples
10000 test samples
Model: "sequential_6"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_14 (Dense)             (None, 128)               100480    
_________________________________________________________________
activation_14 (Activation)   (None, 128)               0         
_________________________________________________________________
dense_15 (Dense)             (None, 128)               16512     
_________________________________________________________________
activation_15 (Activation)   (None, 128)               0         
_________________________________________________________________
dense_16 (Dense)             (None, 10)                1290      
_________________________________________________________________
activation_16 (Activation)   (None, 10)                0         
Total params: 118,282
Trainable params: 118,282
Non-trainable params: 0
_________

48000/48000 [==============================] - 13s 272us/step - loss: 0.1103 - accuracy: 0.9688 - val_loss: 0.1290 - val_accuracy: 0.9653
Epoch 7/50
48000/48000 [==============================] - 12s 248us/step - loss: 0.1083 - accuracy: 0.9697 - val_loss: 0.1285 - val_accuracy: 0.9650
Epoch 8/50
48000/48000 [==============================] - 23s 469us/step - loss: 0.1064 - accuracy: 0.9699 - val_loss: 0.1268 - val_accuracy: 0.9661
Epoch 9/50
48000/48000 [==============================] - 16s 334us/step - loss: 0.1044 - accuracy: 0.9707 - val_loss: 0.1270 - val_accuracy: 0.9648
Epoch 10/50
48000/48000 [==============================] - 17s 346us/step - loss: 0.1029 - accuracy: 0.9714 - val_loss: 0.1245 - val_accuracy: 0.9658
Epoch 11/50
48000/48000 [==============================] - 14s 281us/step - loss: 0.1011 - accuracy: 0.9718 - val_loss: 0.1231 - val_accuracy: 0.9657
Epoch 12/50
48000/48000 [==============================] - 17s 347us/step - loss: 0.0994 - accuracy: 0.9721 - val_l

C:\ProgramData\Anaconda3\lib\site-packages\keras\callbacks\callbacks.py:95: RuntimeWarning: Method (on_train_batch_end) is slow compared to the batch update (0.118482). Check your callbacks.
  % (hook_name, delta_t_median), RuntimeWarning)


48000/48000 [==============================] - 11s 239us/step - loss: 0.0667 - accuracy: 0.9820 - val_loss: 0.1037 - val_accuracy: 0.9713
Epoch 39/50
48000/48000 [==============================] - 11s 230us/step - loss: 0.0658 - accuracy: 0.9825 - val_loss: 0.1040 - val_accuracy: 0.9712
Epoch 40/50
48000/48000 [==============================] - 11s 235us/step - loss: 0.0648 - accuracy: 0.9829 - val_loss: 0.1036 - val_accuracy: 0.9708
Epoch 41/50
48000/48000 [==============================] - 12s 260us/step - loss: 0.0639 - accuracy: 0.9831 - val_loss: 0.1024 - val_accuracy: 0.9712
Epoch 42/50
48000/48000 [==============================] - 11s 223us/step - loss: 0.0630 - accuracy: 0.9832 - val_loss: 0.1018 - val_accuracy: 0.9712
Epoch 43/50
48000/48000 [==============================] - 13s 272us/step - loss: 0.0622 - accuracy: 0.9837 - val_loss: 0.1018 - val_accuracy: 0.9716
Epoch 44/50
48000/48000 [==============================] - 8s 166us/step - loss: 0.0613 - accuracy: 0.9841 - val

<style>
    h1 {
        font-family: "Times New Roman", Times, serif;
        font-size: 30px;
        font-weight: bold;
        text-align: center;
        margin-top: 50px;
        margin-bottom: 50px;
        border-bottom: 3px solid #000;
        text-transform: uppercase;
    }

    p {
        font-family: "Times New Roman", Times, serif;
        font-size: 20px;
        text-align: justify;}
    i {
        font-family: "Times New Roman", Times, serif;
        font-size: 12px;
    }
</style>
        
<h1>Exploring the Impact of Neural Network Parameters on Accuracy Rates</h1>

<p>In this project, I investigated the impact of various parameters on the accuracy rates of a neural network. The MNIST dataset, which is a collection of handwritten digits, was used to train and evaluate the neural network. I focused on three main parameters: the number of hidden layers, batch size, and number of epochs.</p>

<h1>Experiment 1: Number of Hidden Layers</h1>

<p style="padding-bottom: 10px">The number of hidden layers in a neural network is a crucial parameter that determines its complexity. I modified this parameter by testing three different values: <b>1, 5, and 20</b>, and evaluated the accuracy of each model. The table below presents the accuracy rates for each neural network model on the train, validation, and test data:</p> 

<i>The table below summarizes the results of the experiment.</i>

<div align="center">

| Number of Epochs: | 20  | Size of Datasets: | 48000 |Batch Size :|128||
|-------------------|-----|-------------------|-------|-----|-----|-----|
Hidden Layers|Loss   |Accuracy |Validation Loss |Validation Accuracy |Test Score |Test Accuracy 
1            |1.7804 |0.2443   |1.7743          |0.2388              |1.7786     |0.2407
5            |0.6272 |0.8281   |0.6046          |0.8317              |0.6048     |0.8365
20           |0.9737 |0.6314   |0.9557          |0.6442              |0.961      |0.6309

</div>

<p style="padding-top: 20px">As shown in the table, the neural network with <b>5 hidden layers</b> achieved the highest accuracy rates on the validation and test sets, while the neural network with <b>1 hidden layer</b> performed the worst. Interestingly, the neural network with <b>20 hidden layers</b> did not perform as well as the neural network with <b>5 hidden layers</b> in terms of accuracy rate. Although, it did have the lowest validation loss and test score among all the models. This suggests that increasing the number of hidden layers beyond a certain point may not necessarily lead to better performance, and may even result in overfitting.</p>


<p>This experiment highlights the importance of choosing an appropriate number of hidden layers to achieve the best performance in a neural network. It is also important to consider the trade-off between accuracy and overfitting when selecting the number of hidden layers.</p>


<h1>Experiment 2: Exploring the Impact of Batch Size on Accuracy Rates</h1>

<p style="padding-bottom: 20px">In this experiment, I aimed to explore the impact of batch size on the accuracy rates of a neural network. To do this, I trained and evaluated neural networks with different batch sizes on a dataset of handwritten digits (MNIST).</p>

<div align="center">

| Number of Epochs:| 20 | Size of Datasets:| 48000 | Hidden Layers:|128||
|---------|-----------|---------------|---------------|--------------|-----|-----|
Batch Size |Loss   |Accuracy |Validation Loss |Validation Accuracy |Test Score |Test Accuracy 
32         |0.0731 |0.9799   |0.2614          |0.9266              |1.7786     |0.2407
64         |0.0411 |0.9896   |0.0934          |0.9726              |0.0839     |0.9739
256        |0.0345 |0.9921   |0.0916          |0.9720              |0.0809     |0.9751

</div>

<p style="padding-top: 20px">As shown in the table, increasing the batch size from <b>32 to 64</b> improved the accuracy rates on the validation and test sets, as well as reducing the loss on the test set. However, further increasing the batch size from <b>64 to 256</b> did not result in significant improvements in accuracy rates, and only resulted in a slight reduction in loss on the test set. This suggests that using a batch size that is too large may not necessarily lead to better performance.</p>

<p>This experiment highlights the importance of selecting an appropriate batch size for training a neural network, and that the optimal batch size may depend on the specific dataset and model architecture being used.</p>

<h1>Experiment 3: Exploring the Impact of Number of Epochs on Accuracy Rates</h1>

<p style="padding-bottom: 20px">In this experiment, I aimed to explore the impact of the number of epochs on the accuracy rates of a neural network. To do this, I trained and evaluated neural networks with different numbers of epochs on a dataset of handwritten digits (MNIST).</p>

<div align="center">

| Size of Datasets:| 48000 | Hidden Layers:|128|Batch Size: |128||
|---------|-----------|---------------|---------------|--------------|-----|-----|
Number of Epochs |Loss   |Accuracy |Validation Loss |Validation Accuracy |Test Score |Test Accuracy 
10               |0.2710 |0.9228   |0.2534          |0.9284              |0.2528     |0.9276
30               |0.1232 |0.9655   |0.1378          |0.9635              |0.1311     |0.9613
50               |0.0565 |0.9857   |0.0995          |0.9718              |0.0897     |0.9715

</div>

<p style="padding-top:20px">As shown in the table, increasing the number of epochs from <b>10 to 30</b> resulted in significant improvements in accuracy rates on the validation and test sets, as well as reducing the loss on the test set. However, further increasing the number of epochs from <b>30 to 50</b> did not result in significant improvements in accuracy rates, and only resulted in a slight reduction in loss on the test set. This suggests that using too many epochs may lead to overfitting and may not necessarily lead to better performance.</p>

<p>This experiment highlights the importance of selecting an appropriate number of epochs for training a neural network, and that the optimal number of epochs may depend on the specific dataset and model architecture being used.</p>

<h1>Summary of Results Across Experiments</h1>
<p>In my exploration of neural networks, I conducted several experiments to study the impact of different parameters on accuracy rates. Through my experiments, I found that selecting appropriate values for these parameters is essential in achieving optimal performance. Specifically, I discovered that increasing the number of hidden layers can improve accuracy rates up to a certain point, but beyond that point, further increases may lead to overfitting. Additionally, I found that selecting an appropriate batch size is crucial for achieving the best results, as using a batch size that is too large may not necessarily lead to better performance. Finally, I found that increasing the number of epochs can improve accuracy rates, but using too many epochs may lead to overfitting and reduced performance.</p>
<p>These experiments highlight the importance of carefully selecting values for various neural network parameters to achieve the best possible performance. I found that considering both accuracy and overfitting when selecting the number of hidden layers, batch size, and number of epochs is crucial. These findings can be useful for designing and optimizing neural networks for a range of applications.</p>



<h1>References</h1>

- Gulli, A., & Pal, S. (2017). *Deep learning with keras: Get to grips with the basics of keras to implement fast and efficient deep-learning models*. Packt Publishing, Limited.
- Keras. (n.d.). Guides. Retrieved March 6, 2023, from https://keras.io/guides/
- W3Schools. (n.d.). Neural Networks. Retrieved March 06, 2023, from https://www.w3schools.com/ai/ai_neural_networks.asp
racy of the neural network.

Overall, we can see that different parameters have different effects on the accuracy of the neural network. Increasing the number of hidden layers can make the neural network too complex and cause it to overfit the data, while increasing the batch size and number of epochs can improve the accuracy of the neural network up to a certain point. It is important to carefully choose these parameters to optimize the performance of the neural network.